mentat/session.py

```diff
@@ -113,21 +119,51 @@ async def _main(self):
 
         try:
             stream.send("Type 'q' or use Ctrl-C to quit at any time.")
-            stream.send("\nWhat can I do for you?", color="light_blue")
             need_user_request = True
             while True:
                 if need_user_request:
-                    message = await collect_user_input()
+                    # Normally, the code_file_manager pushes the edits; but when agent mode is on, we want all
+                    # edits made between user input to be collected together.
+                    if agent_handler.agent_enabled:
+                        code_file_manager.history.push_edits()
+                        stream.send(
+                            "Use /undo to undo all changes from agent mode since last"
+                            " input.",
+                            color="green",
+                        )
+                    stream.send("\nWhat can I do for you?", color="light_blue")
+                    message = await collect_input_with_commands()
                     if message.data.strip() == "":
                         continue
                     conversation.add_user_message(message.data)
 
-                file_edits = await conversation.get_model_response()
+                parsed_llm_response = await conversation.get_model_response()
                 file_edits = [
-                    file_edit for file_edit in file_edits if file_edit.is_valid()
+                    file_edit
+                    for file_edit in parsed_llm_response.file_edits
+                    if file_edit.is_valid()
                 ]
                 if file_edits:
-                    need_user_request = await get_user_feedback_on_edits(file_edits)
+                    if not agent_handler.agent_enabled:
+                        file_edits, need_user_request = (
+                            await get_user_feedback_on_edits(file_edits)
+                        )
+                    for file_edit in file_edits:
+                        file_edit.resolve_conflicts()
+
+                    applied_edits = await code_file_manager.write_changes_to_files(
+                        file_edits
+                    )
+                    stream.send(
+                        "Changes applied." if applied_edits else "No changes applied.",
+                        color="light_blue",
+                    )
+
+                    if agent_handler.agent_enabled:
+                        if parsed_llm_response.interrupted:
+                            need_user_request = True
```

Comment:
If the user ctrl-c's out, we give them back control. Right now, we still automatically apply changes; I think this makes sense because we're still in 'agent mode', and the user can just /undo, but I can also definitely see wanting it not to apply them (after all, the user interrupted this response). What do y'all think?

Reply:
I think the way you've done it is good. The user already had the chance to interrupt those edits and didn't, and we tell them about `/undo`.